virtwho/config.py

```diff
@@ -1204,23 +1226,44 @@ def init_config(env_options, cli_options):
             continue
         effective_config[section] = new_section
 
-    # Add each section of each file in the drop directory
-    drop_dir_config_sections = effective_config.all_drop_dir_config_sections()
-    for section, values in drop_dir_config_sections.items():
-        new_section = {}
-        new_section.update(**virt_defaults_section)
-        new_section.update(**values)
-        try:
-            new_section = ConfigSection.from_dict(new_section, section, self)
-        except KeyError:
-            # Missing required attribute
-            continue
-        effective_config[section] = new_section
+    effective_config.validate()
+
+    valid_virt_sections = [(name, section) for (name, section) in effective_config.virt_sections()
+                           if section.is_valid()]
+
+    if not valid_virt_sections:
+        has_non_default_env_cli = False
+        validation_errors.append(('warning', 'No valid configurations found'))
+        # Check if ENV_CLI is default, if not fail
+        for name, section in effective_config.items():
+            if name == VW_GLOBAL:
+                # Always keep the global section
+                continue
+            if name == VW_ENV_CLI_SECTION_NAME:
+                has_non_default_env_cli = True
+            validation_errors.append(('warning', 'Dropping invalid configuration "%s"' % name))
+            del effective_config[name]
+        validation_errors.append(('warning',
+                                  'Using default "%s" configuration' % VW_ENV_CLI_SECTION_NAME))
+        # In order to keep compatibility with older releases of virt-who,
+        # fallback to using libvirt as default virt backend
+        # only if we did not have a non_default env/cmdline config
+        if not has_non_default_env_cli:
+            effective_config[VW_ENV_CLI_SECTION_NAME] = ConfigSection.from_dict(
+                    DEFAULTS[VW_ENV_CLI_SECTION_NAME],
+                    VW_ENV_CLI_SECTION_NAME,
+                    effective_config)
```

Comment:
I'm sorry. I meant the block of code between lines 1231-1252 could be refactored into some function (e.g. named `check_valid_conf()`). It is not good practice to have long functions (due to worse readability, etc.). The function `init_config()` is quite long now and it is not worthwhile to make it longer :-).